experimental/Schemes/SpecOpen.jl

```diff
@@ -0,0 +1,675 @@
+  return issubset(complement(intersect(V, ambient(U))), complement(U))
+end
+
+function ==(U::T, V::T) where {T<:SpecOpen}
+  return issubset(U, V) && issubset(V, U)
+end
+
+function ==(
+    U::SpecOpen{BRT, BRET, RT, RET, MST},
+    Y::Spec{BRT, BRET, RT, RET, MST}
+  ) where {BRT, BRET, RT, RET, MST}
+  return issubset(U, Y) && issubset(Y, U)
+end
+
+function ==(
+    Y::Spec{BRT, BRET, RT, RET, MST},
+    U::SpecOpen{BRT, BRET, RT, RET, MST}
+  ) where {BRT, BRET, RT, RET, MST}
+  return U == Y
+end
+
+@Markdown.doc """
+    closure(U::SpecOpen)
+
+Compute the Zariski closure of an open set ``U ⊂ X`` 
+where ``X`` is the affine ambient scheme of ``U``.
+"""
+function closure(U::SpecOpen)
+  X = ambient(U)
+  R = base_ring(OO(X))
+  I = saturated_ideal(localized_modulus(OO(X)))
+  I = saturation(I, ideal(R, gens(U)))
+  return subscheme(X, I)
+end
+
+@Markdown.doc """
+    closure(U::SpecOpen, Y::Spec)
+
+Compute the closure of ``U ⊂ Y``.
+"""
+function closure(
+    U::SpecOpen{BRT, BRET, RT, RET, MST},
+    Y::Spec{BRT, BRET, RT, RET, MST} 
+  ) where {BRT, BRET, RT, RET, MST}
+  issubset(U, Y) || error("the first set is not contained in the second")
+  X = closure(U)
+  return intersect(X, Y)
+end
+
+
+@Markdown.doc """
+    StructureSheafRing{BRT, BRET, RT, RET, MST}
+
+The ring of regular functions ``𝒪(X, U)`` on an open subset ``U`` of an 
+affine scheme ``X``.
+"""
+mutable struct StructureSheafRing{BRT, BRET, RT, RET, MST}
+  scheme::Spec{BRT, BRET, RT, RET, MST}
+  domain::SpecOpen{BRT, BRET, RT, RET, MST}
+
+  function StructureSheafRing(
+      X::Spec{BRT, BRET, RT, RET, MST}, 
+      U::SpecOpen{BRT, BRET, RT, RET, MST}
+    ) where {BRT, BRET, RT, RET, MST}
+    issubset(U, X) || error("open set does not lay in the scheme")
+    return new{BRT, BRET, RT, RET, MST}(X, U)
+  end
+end
+
+@Markdown.doc """
+    scheme(R::StructureSheafRing)
+
+The ring ``R = 𝒪(X, U)`` belongs to a sheaf of rings ``𝒪(X, -)`` and this returns 
+the scheme ``X`` on which ``𝒪`` is defined.
+"""
+scheme(R::StructureSheafRing) = R.scheme
+
+@Markdown.doc """
+    domain(R::StructureSheafRing)
+
+For a ring ``R = 𝒪(X, U)``, return ``U``.
+"""
+domain(R::StructureSheafRing) = R.domain
+
+OO(U::SpecOpen) = StructureSheafRing(ambient(U), U)
+OO(X::Spec{BRT, BRET, RT, RET, MST}, U::SpecOpen{BRT, BRET, RT, RET, MST}) where {BRT, BRET, RT, RET, MST} = StructureSheafRing(X, U)
+
+function ==(R::T, S::T) where {T<:StructureSheafRing} 
+  scheme(R) == scheme(S) || return false
+  domain(S) == domain(R) || return false
+  return true
+end
+
+elem_type(R::StructureSheafRing{BRT, BRET, RT, RET, MST}) where {BRT, BRET, RT, RET, MST}= Type{StructureSheafElem{BRT, BRET, RT, RET, MST}}
+
+@Markdown.doc """
+    StructureSheafElem{BRT, BRET, RT, RET, MST}
+
+An element ``f ∈ 𝒪(X, U)`` of the ring of regular functions on 
+an open set ``U`` of an affine scheme ``X``.
+"""
+mutable struct StructureSheafElem{BRT, BRET, RT, RET, MST}
```

Comment:
Why is this mutable? Can one replace the `domain` or the `restrictions` vector after an instance is created?

Reply:
Writing `mutable` probably just became a habit at this point since usually nothing else works. I will reconsider this.